### PR TITLE
JOSS paper

### PR DESCRIPTION
**Pull request by paraynaud**

### CODE REVIEW

**Comment by amontoison** [comment]
https://github.com/JuliaSmoothOptimizers/KnetNLPModels.jl/actions/runs/5230640348?pr=62

**Comment by paraynaud** [comment]
Was the capital t in "Title" the issue?

**Comment by amontoison** [comment]
> Was the capital t in "Title" the issue?

Yes, It took me 30 minutes to find it...

**Comment by paraynaud** [comment]
Thanks a lot!

**Comment by paraynaud** [comment]
To do list : 
- Move and adapt KnetNLPModel example to Annex
- Add a justification to why we are interested in exploiting quasi-Newton methods (ref)

**Comment by paraynaud** [comment]
What is left to do, e.g. comment not answered by the changes :
- mention : "It is important to stress that we want to be able to experiment with solvers that enforce a decrease of the merit function";
- answer : "Why not just write solvers directly in Flux.jl or Knet.jl";
- mention callback : but it appears in the example section.

After rebasing onto PR-JOSS-Flux and PR-JOSS:
- move Knet model definition to the end to put Flux architecture on the front (with a mention to the Knet model);
- make sure the code calling solver works for both models (same instruction, same callback ...)

Additional:
- delete redundancies during the Summary, the statement of need and the Example (may factorize duplication in both FluxNLPModel or KnetNLPModel example)
- document format (1 line 1 sentence)

**Comment by paraynaud** [comment]
@farhadrclass, @dpo, I made a pass on the document, to fulfill the needs expressed in the previous comment.

Few remarks:
- I'm not convinced by "modularity, flexibility, ecosystem interoperability, and performance optimization" l.41., particularly by "performance optimization".
I don't see how interfaces may optimize the performance, from a software engineering point of view.
For me, the absolute achievement of an interface is to not downgrade the performances of the original module where it comes from.

- I think it is better to have the callback explanation in the numerical example.
Bring the callback in the Summary or in the Statement of need would confuse the reader with implementation details.

Don't hesitate to start a review, I don't see what we can add to this document.

**Comment by farhadrclass** [comment]
@paraynaud 
Thank you for your reviews and edits:
for your remarks:

1. I'm not convinced by "modularity, flexibility, ecosystem interoperability, and performance optimization"...
   1. The idea we are saying is that we decouple the optimization methods and DNN method, hence the modularity
   2. It allow the user to choose what backend they want and what optimization method... flexibility 
   3.  ecosystem interoperability, and performance optimization is the results of modularity approach, allowing the optimization routine and methods to be developed without the need of the DNN models to change 
2.  I agree but @dpo what do you think? 


Other notes:

1.  We need to add @dpo  as co-author 

I can not put review so here are my comments 

- change:

> Once these steps are completed, you can use a solver from JSOSolvers to minimize the loss of LeNetNLPModel. These solvers are originally designed for deterministic optimization. While FluxNLPModel.jl (and KnetNLPModels.jl) manages the loss function to apply it to sampled data, the training minibatch needs to be changed between iterations. This can be achieved using the callback mechanism integrated in JSOSolvers, which executes a predefined function, callback, at the end of each iteration. For more details, refer to the JSOSolvers documentation [@jso].

To

> After completing the necessary steps, one can utilize a solver from JSOSolvers to minimize the loss of LeNetNLPModel. These solvers have been primarily designed for deterministic optimization. In the case of FluxNLPModel.jl (and KnetNLPModels.jl), the loss function is managed to ensure its application to sampled data. However, it is essential to modify the training minibatch between iterations. This objective can be accomplished by leveraging the callback mechanism incorporated in JSOSolvers. This mechanism executes a pre-defined function, known as a callback, at the conclusion of each iteration. For more comprehensive information, please consult the documentation provided by JSOSolvers [@jso].


- change

> Finally, the LeNetNLPModel can be enhanced by using an LSR1 (or LBFGS) approximation of the Hessian, which can be fed to the trunk solver, a quadratic trust-region method with a backtracking line search. To incorporate the LSR1 approximation and trunk into the training process, you can modify the code as follows:


> To enhance the LeNetNLPModel, an LSR1 (or LBFGS) approximation of the Hessian can be employed and fed into the trunk solver. The trunk solver utilizes a quadratic trust-region method with a backtracking line search. To integrate the LSR1 approximation and trunk into the training process, the code can be modified as outlined below:




Overall I am good with it, @dpo  do you have any suggestions ?

>

**Comment by paraynaud** [comment]
Edit l.167

File: paper/paper.md [hunk 1]
```diff
@@ -85,20 +83,6 @@ using FluxNLPModels
 LeNetNLPModel = FluxNLPModel(LeNet, train_loader, test_loader; loss_f = Flux.logitcrossentropy)
 ```
 
-Or to build a Knet.jl architecture, you must define the layers you need: convolutional and dense layers and link them properly.
```
Comment (farhadrclass) [inline paper/paper.md]: are you sure that we don't need to show how KnetNLPModels  is used ?

**Review by farhadrclass — DISMISSED** [review]

File: paper/paper.md (hunk 1 above)
Comment (paraynaud) [reply paper/paper.md]: Yes

**Review by farhadrclass — APPROVED** [review]
